lib/autoupdate/agent/updater.go

```diff
@@ -288,20 +301,143 @@ func (u *Updater) Enable(ctx context.Context, override OverrideConfig) error {
 		}
 	}
 
-	if desiredVersion == "" {
+	if targetVersion == "" {
 		return trace.Errorf("agent version not available from Teleport cluster")
 	}
-	switch cfg.Status.BackupVersion {
-	case "", desiredVersion, cfg.Status.ActiveVersion:
+
+	if err := u.update(ctx, cfg, targetVersion, flags); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Always write the configuration file if enable succeeds.
+
+	cfg.Spec.Enabled = true
+	if err := writeConfig(u.ConfigPath, cfg); err != nil {
+		return trace.Errorf("failed to write %s: %w", updateConfigName, err)
+	}
+	u.Log.InfoContext(ctx, "Configuration updated.")
+	return nil
+}
+
+// Disable disables agent auto-updates.
+// This function is idempotent.
+func (u *Updater) Disable(ctx context.Context) error {
+	cfg, err := readConfig(u.ConfigPath)
+	if err != nil {
+		return trace.Errorf("failed to read %s: %w", updateConfigName, err)
+	}
+	if !cfg.Spec.Enabled {
+		u.Log.InfoContext(ctx, "Automatic updates already disabled.")
+		return nil
+	}
+	cfg.Spec.Enabled = false
+	if err := writeConfig(u.ConfigPath, cfg); err != nil {
+		return trace.Errorf("failed to write %s: %w", updateConfigName, err)
+	}
+	return nil
+}
+
+// Update initiates an agent update.
+// If the update succeeds, the new installed version is marked as active.
+// Otherwise, the auto-updates configuration is not changed.
+// Unlike Enable, Update will not validate or repair the current version.
+// This function is idempotent.
+func (u *Updater) Update(ctx context.Context) error {
+	// Read configuration from update.yaml and override any new values passed as flags.
+	cfg, err := readConfig(u.ConfigPath)
+	if err != nil {
+		return trace.Errorf("failed to read %s: %w", updateConfigName, err)
+	}
+	if err := validateConfigSpec(&cfg.Spec, OverrideConfig{}); err != nil {
+		return trace.Wrap(err)
+	}
+	activeVersion := cfg.Status.ActiveVersion
+	if !cfg.Spec.Enabled {
+		u.Log.InfoContext(ctx, "Automatic updates disabled.", activeVersionKey, activeVersion)
+		return nil
+	}
+	if cfg.Spec.Proxy == "" {
+		return trace.Errorf("Teleport proxy URL must be present in %s", updateConfigName)
+	}
+
+	// Lookup target version from the proxy.
+
+	addr, err := libutils.ParseAddr(cfg.Spec.Proxy)
+	if err != nil {
+		return trace.Errorf("failed to parse proxy server address: %w", err)
+	}
+	resp, err := webclient.Find(&webclient.Config{
+		Context:     ctx,
+		ProxyAddr:   addr.Addr,
+		Insecure:    u.InsecureSkipVerify,
+		Timeout:     30 * time.Second,
+		UpdateGroup: cfg.Spec.Group,
+		Pool:        u.Pool,
+	})
+	if err != nil {
+		return trace.Errorf("failed to request version from proxy: %w", err)
+	}
+	targetVersion := resp.AutoUpdate.AgentVersion
+	var flags InstallFlags
+	if resp.Edition == "ent" {
```

Comment:
as I understand we want isolate import of `modules` package where we have `BuildEnterprise` constant, may be we should add some log error if edition is unknown, might be helpful for debug if by any reason `BuildEnterprise` constant is modified

Reply:
Thanks, good catch, fixed in this PR 🙂